include/zephyr/net/net_l2_connectivity.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for managing L2 layers that have separated post-admin-up connect
+ *	  or associate phases
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+#define ZEPHYR_INCLUDE_NET_L2_CONNECTIVITY_H_
+
+#include <zephyr/device.h>
+#include <zephyr/net/net_if.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief Network Layer 2 connectivity management abstraction layer
+ * @defgroup net_l2_connectivity Network L2 Connectivity Abstraction Layer
+ * @ingroup networking
+ * @{
+ */
+
+/**
+ * @brief Network L2 connectivity API structure
+ *
+ * Used to provide an interface to connect parameters and procedures
+ */
+struct net_l2_connectivity_api {
+	int (*connect)(struct net_if *iface);
+	int (*disconnect)(struct net_if *iface);
+
+	/* Note to reviewers:
+	 * Technically, as long as each iface provides a means by which to set
+	 * association/connection parameters, these get/set_opt funcs are not needed.
+	 *
+	 * However, it can be argued that they should still exist, since they establish a
+	 * standardized pattern for doing that.
+	 *
+	 * The question, then, is: Why limit these to specifically connectivity?
+	 * Should these apply to the whole iface, and be distributed to implementations
+	 * (where they are present) on not only the connectivity_api, but also the l2_api
+	 * and if_api as well?
+	 *
+	 * For now, I have implemented them just on the connectivity_api, but I wonder if they
+	 * should perhaps be more generalized.
+	 */
+	int (*set_opt)(int optname, const void *optval, size_t optlen);
+	int (*get_opt)(int optname, const void *optval, size_t *optlen);
+};
+
+/**
+ * @brief Network interface connectivity structure
+ *
+ * Binds L2 connectivity API and settings to an iface / network device
+ */
+struct net_if_conn {
+	/* The network device the connectivity struct is bound to */
+	struct net_if_dev *if_dev;
+
+	/* The L2 connectivity implementation associated with the network device */
+	struct net_l2_connectivity_api *api;
+
+	/* Per-net-device connectivity settings */
+	int timeout;
+	bool persistence;
+
+	/* To reviewers: Perhaps connectivity settings should be per-l2, rather than per-iface?
```

Comment:
I think we're missing in this picture some kind of connectivity-implementation-specifc context structure, similar to `l2_data` concept for L2. `net_if_conn` could then keep pointers to API and data structures, plus generic parameters, like those above.

Reply:
Do you foresee much headache to add it if we need it?

Reply:
Probably not much of an headache, but I kind of miss the point of having `set_opt()`/`get_opt()` w/o some per-interface data container.

Reply:
I lean towards agreeing with @rlubos on this -- For our purposes, we don't foresee any ifaces sharing connectivity implementations, but since that is something we technically support with this approach, we should probably offer context registration

I'll check out how much work it would be to implement, and if I think I can get it in today, I'll do it. Otherwise, I'll leave it for when we review with a wider audience

Reply:
**Firstly:**

So I am realizing this whole thing is a bit of a miscommunication on my part.

I called `timeout` and `persistence` "connectivity settings", which I meant to be distinct from the settings controlled by `set_opt` and `get_opt`.

So due to my miscommunication, the question I originally intended to ask still stands:

Should timeout, persistence be per-iface, or per-L2?


**Secondly:**

2.) @rlubos You mention a connectivity-implementation-specifc context structure, but then later mention `per-interface data container.`

I think that per-connectivity-implementation data could be stored statically along-side the implementation, so I am not certain there is a need for it.

The set_opt and get_opt functions, would then serve as generic accessors to whatever state the implementation has built for itself, so they would still serve purpose.


**Thirdly:**
I do think you are right, @rlubos -- as-written, there is no way for a connectivity implementation to store per ***netdev/iface*** data.

For this purpose, we should probably provide some kind of context pointer, although I am still trying to work out what I think the best form for it would be.

Reply:
> But storing it per-iface also seems a bit odd, since several ifaces might share the same L2 connection (suppose there are several ifaces assigned to LTE for instance).

IMO **all** connection parameters should be stored at per interface basis. Imagine two Wi-Fi devices, exposing two network interfaces. They may share the connectivity implmentation, but they should have a chance to connect to different networks with different parameters. 

I'm not sure I follow your LTE example. All _real_ network interfaces should have a network device underneath. If we want to have multiple network interfaces working on top of a single device, then we talk about virtual interfaces, but they still need a _real_ interface underneath. In this scenario, the virtual interfaces should not deal with connectivity at all, it should be the responsibility of the underlying interface.

> @rlubos You mention a connectivity-implementation-specifc context structure, but then later mention per-interface data container.

Sorry for not being clear enough. What I meant is that we should follow the L2 data model, where the L2 declares a context structure it expects for the interface to allocate. So taking Ethernet as an example, the Ethernet L2 declare the context structure format it expects to deal with:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/ethernet.h#L520
But every Ethernet interface defines a unique instance of the structure:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/ethernet.h#L844

So to summarize, mutliple interfaces may share the same L2 implementation (L2 API) but they still allocate unique context structures for L2 to deal with for each interface. I think it only makes sense to follow the same pattern for connectivity.

> I do think you are right, @rlubos -- as-written, there is no way for a connectivity implementation to store per netdev/iface data.
For this purpose, we should probably provide some kind of context pointer, although I am still trying to work out what I think the best form for it would be.

I'll again refer to L2 model and `NET_L2_DATA_INIT()` macro. There is some macrology behind it, but the sole purpose of this macro is to allocate a context structure, defined by L2, uniquely for the interface. The structure is then linked with the interface in `NET_IF_INIT()` using `NET_L2_GET_DATA()` to obtain the context structure pointer.

Reply:
>  They may share the connectivity implementation, but they should have a chance to connect to different networks with different parameters.

You're right, a board designer could have, for example, two instances of the same Wi-Fi device model, so it is not accurate for me to assume that if two ifaces share an implementation, they are also sharing a physical device. 

>  All real network interfaces should have a network device underneath. If we want to have multiple network interfaces working on top of a single device, then we talk about virtual interfaces, but they still need a real interface underneath.

My bad, I was operating under the misunderstanding that non-virtual network devices could share physical devices. If this is disallowed, then storing timeout and persistence per-iface seems perfectly reasonable.


These two corrections in mind, I completely agree with your stance. Fortunately, I speculatively worked on an implementation yesterday assuming I would wind up agreeing, so most of this is already in place. I will try to have it merged some time today.